src/main/scala/mesosphere/marathon/state/PortDefinition.scala

```diff
@@ -0,0 +1,48 @@
+package mesosphere.marathon.state
+
+import com.wix.accord._
+import com.wix.accord.dsl._
+import com.wix.accord.{ Failure, RuleViolation, Success }
+import mesosphere.marathon.api.v2.Validation._
+import org.slf4j.LoggerFactory
+
+import scala.collection.immutable.Seq
+
+case class PortDefinition(port: Int,
+                          protocol: String = "tcp",
+                          name: Option[String] = None,
+                          labels: Map[String, String] = Map.empty[String, String])
+
+object PortDefinition {
+  implicit val portDefinitionValidator = validator[PortDefinition] { portDefinition =>
+    portDefinition.protocol is oneOf(DiscoveryInfo.Port.AllowedProtocols)
+    portDefinition.port should be >= 0
+  }
+}
+
+object PortDefinitions {
+  def apply(ports: Int*): Seq[PortDefinition] = {
+    ports.map(PortDefinition.apply(_)).toIndexedSeq
+  }
+
+  private[this] def uniquePorts: Validator[Seq[PortDefinition]] = new Validator[Seq[PortDefinition]] {
+    override def apply(portDefinitions: Seq[PortDefinition]): Result = {
+      val nonRandomPorts = portDefinitions.map(_.port).filterNot(_ == AppDefinition.RandomPortValue)
+
+      if (nonRandomPorts.size == nonRandomPorts.distinct.size) Success
+      else Failure(Set(RuleViolation("portDefinitions", "Ports must be unique.", None)))
+    }
+  }
+
+  private[this] def uniquePortNames: Validator[Seq[PortDefinition]] = new Validator[Seq[PortDefinition]] {
```

Comment:
`Validator.elementsAreUnique` only accepts a "filter" function that filter/maps from Seq[A] to Seq[A].

In this case we need to map from `Seq[PortDefinitions]` to `Seq[Int]` and then apply the validation logic.

That means that I can not use `Validator.elementsAreUnique` as-is. I'll try to refactor it later.